odl/space/fspace.py

```diff
@@ -154,199 +182,293 @@ class FunctionSetVector(Operator):
 
-            else:
-                raise TypeError('input is neither an element of the '
-                                'function domain {!r} nor an array or '
-                                'meshgrid-type coordinate list.'
-                                ''.format(self.domain))
+        Raises
+        ------
+        TypeError
+            If `x` is not a valid vectorized evaluation argument
+
+            If `out` is not a range element or a `numpy.ndarray`
+            of range elements
 
-            if (min_coords not in self.domain or
-                    max_coords not in self.domain):
+        ValueError
+            If evaluation points fall outside the valid domain
+        """
+        vec_bounds_check = kwargs.pop('vec_bounds_check', True)
+        if vec_bounds_check and not hasattr(self.domain, 'contains_all'):
+            raise AttributeError('vectorized bounds check not possible for '
+                                 'domain {}, missing `contains_all()` '
+                                 'method.'.format(self.domain))
+
+        # Check for input type and determine output shape
+        if is_valid_input_array(x, self.domain.ndim):
+            out_shape = out_shape_from_array(x)
+            scalar_out = False
+            # For 1d, squeeze the array
+            if self.domain.ndim == 1 and x.ndim == 2:
+                x = x[0]
+        elif is_valid_input_meshgrid(x, self.domain.ndim):
+            out_shape = out_shape_from_meshgrid(x)
+            scalar_out = False
+            # For 1d, fish out the vector from the tuple
+            if self.domain.ndim == 1:
+                x = x[0]
+        elif x in self.domain:
+            x = np.atleast_2d(x).T  # make a (d, 1) array
+            out_shape = (1,)
+            scalar_out = (out is None)
+        else:
+            # Unknown input
+            raise TypeError('argument {!r} not a valid vectorized '
+                            'input. Expected an element of the domain '
+                            '{dom}, a ({dom.ndim}, n) array '
+                            'or a length-{dom.ndim} meshgrid sequence.'
+                            ''.format(x, dom=self.domain))
+
+        # Check bounds if specified
+        if vec_bounds_check:
+            if not self.domain.contains_all(x):
                 raise ValueError('input contains points outside '
-                                 '`domain` {}.'.format(self.domain))
+                                 'the domain {}.'.format(self.domain))
+
+        # Call the function and check out shape, before or after
+        if out is None:
+            out = self._call(x, **kwargs)
+            if out_shape != (1,) and out.shape != out_shape:
+                raise ValueError('output shape {} not equal to shape '
+                                 '{} expected from input.'
+                                 ''.format(out.shape, out_shape))
+        else:
+            if not isinstance(out, np.ndarray):
```

Comment:
Would be possible in the case of a `FunctionSpace`, but it's harder for a general `FunctionSet`. I'd say if a user provides a messed up data type array, it's not our fault if the data does not end up to be the way it should.